### PR TITLE
Ra dspdc 1152 publish diffs

### PR DESCRIPTION
**Pull request by raaidbroad**
Okay so this adds two steps to the end-to-end workflow:

1. A step that computes the diffs between two ClinVAR releases and writes them to a bucket (created, updated, and deleted rows)
2. A step that pings the clingen kafka topic in the spec that they want (written out below for clarity here)

The message sent to their kafka topic is of the following format:

```
{
  "release_date": "2019-04-03",
  "bucket": "broad-dsp-monster-clingen-prod-ingest-results",
  "files": [
    "backfill_2019_04/clinical_assertion/created/000000000000",
    "backfill_2019_04/clinical_assertion/created/000000000001",
    "backfill_2019_04/clinical_assertion_observation/created/000000000000",
    "backfill_2019_04/clinical_assertion_trait/created/000000000000"
  ]
}
```

This also adds a docker image definition that is used in the diffing steps.

### CODE REVIEW

File: docker/diff/diff.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+dest_client = bigquery.Client(project=dest_data_project) if dest_data_project else None
+
+
+class DiffType(Enum):
+    CREATE = "created"
+    DELETE = "deleted"
+    UPDATE = "updated"
+
+
+def create_dataset(dataset_name: str = dest_dataset_name, client: bigquery.Client = dest_client) -> str:
```
Comment (pshapiro4broad) [inline docker/diff/diff.py:41]: This doesn't appear to be used.

File: docker/diff/diff.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        join = "RIGHT JOIN"
+        on = "ON A.id = B.id"
+        where = "WHERE A.id IS NULL"
+    elif diff_type is DiffType.UPDATE:
+        select_table = "A"
+        join = "JOIN"
+        # for update, we need everything where the ids are the same but ANYTHING at all in the other columns is
+        # different. We join on the id, AND we join on inequality between all other columns, so all the other cols
+        # are OR-ed together. Some of them are ARRAY cols, so we need to convert them to string before comparing.
+        diff_join = " OR ".join([
```
Comment (pshapiro4broad) [inline docker/diff/diff.py:139]: wow

File: docker/diff/diff.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            for col_name, col_type in cols.items()])
+        on = f"ON A.id = B.id AND ({diff_join})"
+        where = ""
+    else:
+        raise ValueError("Invalid DiffType for get_rows")
+
+    selected_cols = ", ".join([f"{select_table}.{col}" for col in tuple(cols)])
+    query = f"""
+    SELECT {select_table}.id, {selected_cols} FROM
+    (SELECT * FROM `{repo_data_project}.{repo_dataset_name}.{table_name}` WHERE {version_col} = "{release_date}") AS A
```
Comment (pshapiro4broad) [inline docker/diff/diff.py:151]: I'm not sure about the back tick vs double quotes here... do these quotes have different meanings in BQ? In SQL you usually see single quote `'` for all SQL string literals.

File: docker/diff/diff.py (hunk 1 above)
Comment (raaidbroad) [reply docker/diff/diff.py:41]: Correct; I'd like to have this function or something similar be the way we create datasets moving forward (I thought I'd need to create one but I can piggyback off of a previous step's dataset creation). I can remove it and store it somewhere else for now if that makes more sense.

File: docker/diff/diff.py (hunk 2 above)
Comment (raaidbroad) [reply docker/diff/diff.py:139]: yeppp

File: docker/diff/diff.py (hunk 3 above)
Comment (raaidbroad) [reply docker/diff/diff.py:151]: They do; the back ticks are specifically for writing out the whole project-name.dataset-name.table-name thing I believe, it's a special difference in their "standard SQL"

File: docker/diff/diff.py (hunk 1 above)
Comment (pshapiro4broad) [reply docker/diff/diff.py:41]: OK to leave it here, I just wanted to make sure I was reading things right.

**Pull request merged**